### PR TITLE
Drop support for EOL Python 2.7

### PR DESCRIPTION
**Pull request by hugovk**
Drop support for ~legacy~ EOL Python 2.7, fixes https://github.com/NeuralEnsemble/python-neo/issues/699.

Python 3.4 is EOL and no longer receiving security updates (or any updates) from the core Python team.

Version | Release date | Supported until
--- | ---------- | ----------
2.7 | 2010-07-03 | 2020-01-01
3.4 | 2014-03-16 | 2019-03-16

Source: https://en.wikipedia.org/wiki/CPython#Version_history

It's also little used. Here's the pip installs for neo from PyPI for June 2019:

| category | percent | downloads |
|----------|--------:|----------:|
| 3.6      |  55.89% |     2,994 |
| 3.7      |  24.85% |     1,331 |
| 2.7      |  14.11% |       756 |
| 3.5      |   3.60% |       193 |
| null     |   1.31% |        70 |
| 3.4      |   0.22% |        12 |
| 3.3      |   0.02% |         1 |
| Total    |         |     5,357 |

Source: `pypistats python_minor neo --last-month # pip install pypistats`

### CODE REVIEW

**Comment by apdavison** [comment]
Many thanks for this. It's very interesting to see the stats.

Our plan was to wait until Python 2.7 reaches EOL before dropping support, so we'll merge this PR then.

**Comment by hugovk** [comment]
You're welcome! I've made PR #729 to remove the already EOL Python 3.4 and do the upgrade parts that are already possible now.

**Comment by hugovk** [comment]
Resolved conflicts and updated title. Happy new year!

**Comment by JuliaSprenger** [comment]
@hugovk Thanks again for this PR! I am looking into it since we are ready now the including these changes. Unfortunately there appeared a couple of merge conflicts. Would you have time to fix these so we can continue here?

File: neo/rawio/bci2000rawio.py [hunk 1]
```diff
@@ -248,7 +246,7 @@ def parse_dimensions(param_list):
         return num_els, el_labels
 
     import io
```
Comment (JuliaSprenger) [inline neo/rawio/bci2000rawio.py:248]: This import is also not needed anymore.

File: neo/rawio/brainvisionrawio.py [hunk 2]
```diff
@@ -186,7 +184,7 @@ def _rescale_event_timestamp(self, event_timestamps, dtype):
 
 
 def read_brainvsion_soup(filename):
-    with io.open(filename, 'r', encoding='utf8') as f:
+    with open(filename, 'r', encoding='utf8') as f:
```
Comment (JuliaSprenger) [inline neo/rawio/brainvisionrawio.py:187]: Please also remove the import of `io`, since this is not needed anymore.

**Comment by hugovk** [comment]
Unused `io` imports removed.

**Comment by JuliaSprenger** [comment]
@samuelgarcia @apdavison Ready to merge from my side. Any more comments from you?

**Comment by apdavison** [comment]
| category | percent | downloads |
|----------|--------:|----------:|
| 3.7      |  42.47% |     4,806 |
| 3.5      |  30.27% |     3,425 |
| 3.6      |  11.63% |     1,316 |
| 3.8      |  10.82% |     1,224 |
| 2.7      |   3.61% |       408 |
| null     |   1.18% |       134 |
| 3.4      |   0.03% |         3 |
| Total    |         |    11,316 |
Latest figures from pypistats. I think these justify dropping 2.7. I will review the PR as soon as I have time.

File: neo/io/blackrockio_v4.py [hunk 3]
```diff
@@ -386,7 +384,7 @@ def __extract_nsx_file_spec(self, nsx_nb):
             spec = '{}.{}'.format(
                 nsx_file_id['ver_major'], nsx_file_id['ver_minor'])
         else:
-            raise IOError('Unsupported NSX file type.')
+            raise OSError('Unsupported NSX file type.')
```
Comment (apdavison) [inline neo/io/blackrockio_v4.py:387]: This should not be an `OSError`. I suggest leaving it as `IOError` for now, then we can replace it with something suitable, e.g. a custom `NotSupportedError`, in future.

File: neo/io/blackrockio_v4.py [hunk 4]
```diff
@@ -407,7 +405,7 @@ def __extract_nev_file_spec(self):
             spec = '{}.{}'.format(
                 nev_file_id['ver_major'], nev_file_id['ver_minor'])
         else:
-            raise IOError('NEV file type {} is not supported'.format(
+            raise OSError('NEV file type {} is not supported'.format(
```
Comment (apdavison) [inline neo/io/blackrockio_v4.py:408]: same comment as above

File: neo/io/igorproio.py [hunk 5]
```diff
@@ -101,7 +99,7 @@ def read_analogsignal(self, path=None, lazy=False):
             data = bw.load(self.filename)
             version = data['version']
             if version > 5:
-                raise IOError("Igor binary wave file format version {} "
+                raise OSError("Igor binary wave file format version {} "
```
Comment (apdavison) [inline neo/io/igorproio.py:102]: same as previous comment. Keep as IOError for now, to be replaced with a custom exception in a future PR

File: neo/rawio/axonrawio.py [hunk 6]
```diff
@@ -317,7 +313,7 @@ def read_raw_protocol(self):
         info = self._axon_info
 
         if info['fFileVersionNumber'] < 2.:
-            raise IOError("Protocol section is only present in ABF2 files.")
+            raise OSError("Protocol section is only present in ABF2 files.")
```
Comment (apdavison) [inline neo/rawio/axonrawio.py:316]: keep as IOError for now

File: neo/rawio/blackrockrawio.py [hunk 7]
```diff
@@ -799,7 +797,7 @@ def __extract_nsx_file_spec(self, nsx_nb):
             spec = '{}.{}'.format(
                 nsx_file_id['ver_major'], nsx_file_id['ver_minor'])
         else:
-            raise IOError('Unsupported NSX file type.')
+            raise OSError('Unsupported NSX file type.')
```
Comment (apdavison) [inline neo/rawio/blackrockrawio.py:800]: keep as IOError for now

File: neo/rawio/blackrockrawio.py [hunk 8]
```diff
@@ -820,7 +818,7 @@ def __extract_nev_file_spec(self):
             spec = '{}.{}'.format(
                 nev_file_id['ver_major'], nev_file_id['ver_minor'])
         else:
-            raise IOError('NEV file type {} is not supported'.format(
+            raise OSError('NEV file type {} is not supported'.format(
```
Comment (apdavison) [inline neo/rawio/blackrockrawio.py:821]: keep as  IOError for now

File: neo/rawio/tdtrawio.py [hunk 9]
```diff
@@ -128,7 +124,7 @@ def _parse_header(self):
                             break
                 except OSError:
                     pass
-                except IOError:
+                except OSError:
```
Comment (apdavison) [inline neo/rawio/tdtrawio.py:127]: this now duplicates the previous two lines

File: neo/test/iotest/test_stimfitio.py [hunk 10]
```diff
@@ -1,33 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (apdavison) [inline neo/test/iotest/test_stimfitio.py:1]: I don't think we should remove this unless there are no plans to make stimfitio work with Python 3. The skipped tests are a reminder to fix this.

**Comment by apdavison** [comment]
@samuelgarcia ok to drop Python 2.7?

File: neo/io/blackrockio_v4.py [hunk 11]
```diff
@@ -271,7 +269,7 @@ def __init__(self, filename, nsx_override=None, nev_override=None,
 
         # check if there are any files present
         if not any(list(self._avail_files.values())):
-            raise IOError(
+            raise OSError(
```
Comment (JuliaSprenger) [inline neo/io/blackrockio_v4.py:272]: This should then also stay an `IOError`

File: neo/rawio/blackrockrawio.py [hunk 12]
```diff
@@ -173,7 +171,7 @@ def __init__(self, filename=None, nsx_override=None, nev_override=None,
                     self._avail_nsx.append(int(ext[-1]))
 
         if not self._avail_files['nev'] and not self._avail_nsx:
-            raise IOError("No Blackrock files found in specified path")
+            raise OSError("No Blackrock files found in specified path")
```
Comment (JuliaSprenger) [inline neo/rawio/blackrockrawio.py:174]: This should also stay an `IOError`.

File: neo/rawio/blackrockrawio.py (hunk 8 above)
Comment (JuliaSprenger) [inline neo/rawio/blackrockrawio.py:821]: This should also stay an `IOError`.

**Pull request merged**